### PR TITLE
CHE-4144: Recognize when nodejs is froze and stop process

### PR DESCRIPTION
**Pull request by tolusha**
**Problem**
NodeJs has problem with debugging arbitrary applications.  When application is finished nodejs console is froze.

**Solution**
Continuously execute some command to check if nodejs proceed output.
If no more output after some period of time then debugger will be ended.

### What does this PR do?
Recognize when application is finished and ends debugger.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/4144

#### Changelog
Fix situations where node.js debugger doesn't stop and can become frozen.

#### Release Notes
Not required

#### Docs PR
Not required

### CODE REVIEW

**Review by AndrienkoAleksandr — APPROVED** [review]

File: plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/NodeJsDebugProcess.java [hunk 1]
```diff
@@ -107,115 +99,79 @@ void stop() {
...
                 if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
-                    LOG.warn("Unable to terminate main pool");
+                    LOG.warn("Unable to terminate main pool.");
                 }
             }
         } catch (InterruptedException e) {
-            LOG.warn(e.getMessage());
+            if (!executor.isShutdown()) {
+                LOG.warn("Unable to terminate main pool.");
+            }
```
Comment (mmorhun) [inline plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/NodeJsDebugProcess.java]: Please handle `InterruptedException` and set interrupted flag. Why after interrupt you wait up to 10 seconds below?

File: plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/NodeJsDebugProcess.java [hunk 2]
```diff
@@ -32,70 +30,64 @@
...
-        executor.scheduleWithFixedDelay(new OutputReader(), 0, 100, TimeUnit.MILLISECONDS);
+        executor = Executors
+                .newScheduledThreadPool(1, new ThreadFactoryBuilder()
+                        .setNameFormat("nodejs-debugger-%d")
+                        .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                        .setDaemon(true)
+                        .build());
+
+        OutputReader outputReader = new OutputReader(process, outputSeparator, this::notifyObservers);
+        executor.scheduleWithFixedDelay(outputReader, 0, 100, TimeUnit.MILLISECONDS);
```
Comment (skabashnyuk) [inline plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/NodeJsDebugProcess.java:64]: Why don't you use Schedule annotation here?

File: plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/OutputReader.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Continuously reads process output and store in the {@code #outputs}.
```
Comment (skabashnyuk) [inline plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/OutputReader.java:21]: author

**Review by bmicklea — APPROVED** [review]

File: plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/NodeJsDebugProcess.java (hunk 2 above)
Comment (tolusha) [reply plugins/plugin-nodejs-debugger/che-plugin-nodejs-debugger-server/src/main/java/org/eclipse/che/plugin/nodejsdbg/server/NodeJsDebugProcess.java:64]: `OutputReader` is not rely on container lifecycle.
It is started and stopped on demand.

**Review by mmorhun — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**